File: src/go/pkg/netipc/service/raw/increment.go
package raw

import "github.com/netdata/netdata/go/plugins/pkg/netipc/protocol"

// IncrementHandler serves a single INCREMENT service kind.
type IncrementHandler func(uint64) (uint64, bool)

// CallIncrement performs a blocking INCREMENT call.
// Sends requestValue, returns the server's response value.
func (c *Client) CallIncrement(requestValue uint64) (uint64, error) {
	return c.CallIncrementWithTimeout(requestValue, 0)
}

func (c *Client) CallIncrementWithTimeout(requestValue uint64, timeoutMs uint32) (uint64, error) {
	if err := c.validateMethod(protocol.MethodIncrement); err != nil {
		return 0, err
	}

	var result uint64

	err := c.callWithRetry(func() error {
		var reqBuf [protocol.IncrementPayloadSize]byte
		if protocol.IncrementEncode(requestValue, reqBuf[:]) == 0 {
			return protocol.ErrTruncated
		}

		_, payload, rerr := c.doRawCallWithTimeout(protocol.MethodIncrement, reqBuf[:], timeoutMs)
		if rerr != nil {
			return rerr
		}

		val, derr := protocol.IncrementDecode(payload)
		if derr != nil {
			return derr
		}
		result = val
		return nil
	})
	return result, err
}

// CallIncrementBatch performs a blocking batch INCREMENT call.
// Sends multiple values, returns the server's response values.
func (c *Client) CallIncrementBatch(values []uint64) ([]uint64, error) {
	return c.CallIncrementBatchWithTimeout(values, 0)
}

func (c *Client) CallIncrementBatchWithTimeout(values []uint64, timeoutMs uint32) ([]uint64, error) {
	if err := c.validateMethod(protocol.MethodIncrement); err != nil {
		return nil, err
	}

	if len(values) == 0 {
		return nil, nil
	}

	var results []uint64
	itemCount, err := checkedLookupU32(len(values))
	if err != nil {
		return nil, err
	}

	err = c.callWithRetry(func() error {
		batchBufSize, err := incrementBatchRequestSize(len(values))
		if err != nil {
			return err
		}
		batchBuf := ensureClientScratch(&c.requestBuf, batchBufSize)
		bb := protocol.NewBatchBuilder(batchBuf, itemCount)

		for _, v := range values {
			var item [protocol.IncrementPayloadSize]byte
			if protocol.IncrementEncode(v, item[:]) == 0 {
				return protocol.ErrTruncated
			}
			if err := bb.Add(item[:]); err != nil {
				return err
			}
		}

		totalPayloadLen, _ := bb.Finish()
		reqPayload := batchBuf[:totalPayloadLen]

		hdr := protocol.Header{
			Kind:            protocol.KindRequest,
			Code:            protocol.MethodIncrement,
			Flags:           protocol.FlagBatch,
			ItemCount:       itemCount,
			MessageID:       uint64(c.callCount) + 1,
			TransportStatus: protocol.StatusOK,
		}

		if err := c.transportSend(&hdr, reqPayload); err != nil {
			return err
		}

		respHdr, respPayload, err := c.transportReceiveWithControl(
			c.resolvedCallTimeout(timeoutMs), c.abortSignal())
		if err != nil {
			return err
		}

		if respHdr.Kind != protocol.KindResponse {
			return protocol.ErrBadKind
		}
		if respHdr.Code != protocol.MethodIncrement {
			return protocol.ErrBadLayout
		}
		if respHdr.MessageID != hdr.MessageID {
			return protocol.ErrBadLayout
		}
		switch respHdr.TransportStatus {
		case protocol.StatusOK:
		case protocol.StatusLimitExceeded:
			if current := c.sessionMaxResponsePayloadBytes(); current > 0 {
				if current >= ^uint32(0)/2 {
					c.noteResponseCapacity(^uint32(0))
				} else {
					c.noteResponseCapacity(current * 2)
				}
			}
			return protocol.ErrOverflow
		default:
			return protocol.ErrBadLayout
		}
		if respHdr.Flags&protocol.FlagBatch == 0 || respHdr.ItemCount != itemCount {
			return protocol.ErrBadItemCount
		}

		out := make([]uint64, itemCount)
		for i := range itemCount {
			itemData, gerr := protocol.BatchItemGet(respPayload, itemCount, i)
			if gerr != nil {
				return gerr
			}
			val, derr := protocol.IncrementDecode(itemData)
			if derr != nil {
				return derr
			}
			out[i] = val
		}
		results = out
		return nil
	})
	return results, err
}

func incrementBatchRequestSize(valueCount int) (int, error) {
	if _, err := checkedLookupU32(valueCount); err != nil {
		return 0, err
	}
	dirBytes, err := checkedLookupMul(valueCount, 8)
	if err != nil {
		return 0, err
	}
	dirAligned, err := checkedLookupAlign8(dirBytes)
	if err != nil {
		return 0, err
	}
	itemsBytes, err := checkedLookupMul(valueCount, protocol.IncrementPayloadSize)
	if err != nil {
		return 0, err
	}
	paddingBytes, err := checkedLookupMul(valueCount, protocol.Alignment)
	if err != nil {
		return 0, err
	}
	batchBufSize, err := checkedLookupAdd(dirAligned, itemsBytes)
	if err != nil {
		return 0, err
	}
	return checkedLookupAdd(batchBufSize, paddingBytes)
}

// IncrementDispatch adapts a typed increment handler to the raw dispatch shape.
func IncrementDispatch(handle IncrementHandler) DispatchHandler {
	if handle == nil {
		return nil
	}
	return func(request []byte, responseBuf []byte) (int, error) {
		value, err := protocol.IncrementDecode(request)
		if err != nil {
			return 0, err
		}
		result, ok := handle(value)
		if !ok {
			return 0, errHandlerFailed
		}
		n := protocol.IncrementEncode(result, responseBuf)
		if n == 0 {
			return 0, protocol.ErrOverflow
		}
		return n, nil
	}
}

File: src/crates/netipc/src/service/raw/increment.rs
use super::client::{ClientConfig, RawCallKind, RawClient};
use super::dispatch::{DispatchError, DispatchHandler};
use crate::protocol::{
    self, batch_item_get, increment_decode, increment_encode, BatchBuilder, NipcError,
    INCREMENT_PAYLOAD_SIZE, METHOD_INCREMENT,
};
use std::sync::Arc;

pub type IncrementHandler = Arc<dyn Fn(u64) -> Option<u64> + Send + Sync>;

impl RawClient {
    /// Create a new client context bound to the increment service kind.
    /// Does NOT connect. Does NOT require the server to be running.
    pub fn new_increment(run_dir: &str, service_name: &str, config: ClientConfig) -> Self {
        Self::new_bound(run_dir, service_name, METHOD_INCREMENT, config)
    }

    /// Blocking typed call: INCREMENT method.
    /// Sends a u64 value, receives the incremented u64 back.
    pub fn call_increment(&mut self, value: u64) -> Result<u64, NipcError> {
        self.validate_method(METHOD_INCREMENT)?;
        let req_len = {
            let req_buf = self.request_scratch(INCREMENT_PAYLOAD_SIZE);
            let req_len = increment_encode(value, req_buf);
            if req_len == 0 {
                return Err(NipcError::Truncated);
            }
            req_len
        };

        let response =
            self.raw_call_with_retry(METHOD_INCREMENT, req_len, RawCallKind::single())?;
        increment_decode(self.response_payload(response)?)
    }

    /// Blocking typed batch call: INCREMENT method.
    /// Sends multiple u64 values, receives the incremented u64s back.
    pub fn call_increment_batch(&mut self, values: &[u64]) -> Result<Vec<u64>, NipcError> {
        self.validate_method(METHOD_INCREMENT)?;
        if values.is_empty() {
            return Ok(Vec::new());
        }

        if values.len() == 1 {
            let r = self.call_increment(values[0])?;
            return Ok(vec![r]);
        }

        let count = values.len() as u32;
        let req_buf_size = protocol::align8(count as usize * 8)
            + count as usize * protocol::align8(INCREMENT_PAYLOAD_SIZE)
            + 64;
        let req_len = {
            let req_buf = self.request_scratch(req_buf_size);
            let mut bb = BatchBuilder::new(req_buf, count);
            for &v in values {
                let mut item_buf = [0u8; INCREMENT_PAYLOAD_SIZE];
                if increment_encode(v, &mut item_buf) == 0 {
                    return Err(NipcError::Truncated);
                }
                bb.add(&item_buf).map_err(|_| NipcError::Overflow)?;
            }
            let (req_len, _out_count) = bb.finish();
            req_len
        };

        let response =
            self.raw_call_with_retry(METHOD_INCREMENT, req_len, RawCallKind::batch(count))?;
        let resp_payload = self.response_payload(response)?;
        let mut results = Vec::with_capacity(values.len());
        for i in 0..count {
            let (item_data, _item_len) = batch_item_get(resp_payload, count, i)?;
            let val = increment_decode(item_data)?;
            results.push(val);
        }

        Ok(results)
    }
}

pub fn increment_dispatch(handler: IncrementHandler) -> DispatchHandler {
    Arc::new(move |request, response_buf| {
        let value = increment_decode(request).map_err(|_| DispatchError::BadEnvelope)?;
        let result = handler(value).ok_or(DispatchError::HandlerFailed)?;
        let n = increment_encode(result, response_buf);
        if n == 0 {
            return Err(DispatchError::Overflow);
        }
        Ok(n)
    })
}
